src/main/java/org/mockito/internal/creation/bytebuddy/SubclassBytecodeGenerator.java

```diff
@@ -97,10 +98,31 @@ private static boolean needsSamePackageClassLoader(MockFeatures<?> features) {
             // package private methods.
             return true;
         }
+        if (hasPackagePrivateParam(features.mockedType)) {
```

Comment:
I wonder what the performance implications of this change are. Depending on the amount of methods declared by the class or any of its parents, this could potentially lead to a considerable amount of checks. Could you do a quick check if we are regressing in terms of performance (I think we can mock a class with loads of methods and see what the characteristics are)?

If the performance impact is larger than ideal, we could consider caching the results with a `org.mockito.internal.util.concurrent.WeakConcurrentSet`

Reply:
I thought about that, but checking modifiers is a JVM intrinsic backed by a single word memory fetch. If I was recursively checking the class tree you'd be right to worry about pathological interface DAGs causing exponential blowup, but that's not happening here.

In addition, we short-circuit on per-type checks before we check methods, and we are only checking public methods. So this path only gets hit if the mocked type is a public interface, and it only checks public methods. By definition the methods are unique for a single class, so there's no benefit to memoizing at that level. And the per-param modifier checks are much cheaper than doing a hashmap store/lookup.

And all in all, every single method and param we process this way has to get parsed and generated in the actual mock class anyway, which is inherently more expensive. So I'm not concerned about performance here - if you mock an interface with tens of thousands of methods, we have to generate a class with tens of thousands of methods, verify it, and load it, which will always be more expensive.

Reply:
Thanks for the explanation! Yes that makes sense to me 😄